### PR TITLE
refactor path handling

### PR DESCRIPTION
**Pull request by flatcap**
### About

Encapsulate Mailbox Paths:

```c
struct Path
{
  char *desc;      ///< Descriptive name    
  char *orig;      ///< User-entered path    
  char *canon;     ///< Canonical path    
  int type;        ///< Path type, enum MailboxType    
  PathFlags flags; ///< Flags describing what's known about the path    
};

```

Starting from an 'unknown' state, such as 'user entry', a Mailbox Path can undergo several transformations:

- **Resolve**
  Expand '+' to `$folder`, '!' to the spoolfile, `@alias` to a username
  This is performed by NeoMutt

- **Tidy**
  Light tidying of the path, e.g. fixing the case; removing duplicate path separators
  This is performed by the backend

- **Canon**
  Create the canonical path, e.g. follow symlinks; sort parameters
  This is performed by the backend

### Done

- mxapi: tidy stat(2) usage
  Add `bool MxOps.is_local` so NeoMutt knows when `stat(2)` is appropriate

- config: add path type
  Encapsulate a Mailbox 'path'
  
- test: add config/path
  Test coverage for 'path' config type

- libmutt: split tilde from canon
  Expand `~` before saving into a config variable
  This means we don't need to pre-process a variable before calling `fopen()`
  
- config: convert path variables
  Change 24 config variables to `DT_PATH` (using `Path`)
  Each variable is subtyped as either `DT_PATH_DIR` or `DT_PATH_FILE`

- mxapi: add new path functions
  Add stubs for all the new path functions
  Each backend now has a `path.c` file which shouldn't have any dependencies

### To Do

- Fill out the path functions (mostly cut'n'paste from the existing ones)
- Decide what a 'tidy' path means for each backend
- Decide what a 'canonical' path means for each backend
- Replace each `char` array, representing a Mailbox, with `Path`
- Strip out old MxOps path functions
- Rename new path functions
- Strip out redundant backend path functions

### CODE REVIEW

**Comment by flatcap** [comment]
**Local filesystem paths**:

- file | dir
- absolute | relative | tilde
- real | symlink
- clean | dirty
- trailing slash | none

## PROBE - Tests for matching a Mailbox Path

**compressed**

- path matches `open-hook` regex
- existence of dir isn't tested
- existence of file isn't tested

**imap**

- path begins `imap://` or `imaps://` (ignore case)
- URI of form: `imap://user:pass@host.com:port/path/path`

**maildir**

- path must be a directory
- `dir` must exist
- `dir/cur` must exist
- `dir/new`, `dir/tmp` aren't tested

**mbox**

- path must be a file
- `file` must exist
- `file` can be empty
- `file` must begin: `From ` (leading blank lines are ignored)

**mmdf**

- path must be a file
- `file` must exist
- `file` can be empty
- `file` must begin: `From ` (leading blank lines are ignored)

**mh**

- path must be a directory
- `dir` must exist
- subdir must exist, one of:
  - `.mh_sequences`
  - `.xmhcache`
  - `.mew_cache`
  - `.mew-cache`
  - `.sylpheed_cache`
  - `.overview`

**nntp**

- path begins `news://` or `snews://` (ignore case)
- URI of form: `news://user:pass@host.com:port/path/path`

**notmuch**

- path begins `notmuch://` (ignore case)
- URI of form `notmuch:///path/to/db?query&param&param`

**pop**

- path begins `pop://` or `pops://` (ignore case)
- URI of form: `pop://user:pass@host.com:port/`

## TIDY - Rules for tidying a Mailbox Path

**Local filesystem paths**:

- tilde
  - expand `~`, `~/`, `~user`
  - then follow absolute rules
- relative
  - expand to absolute (relative to NeoMutt's cwd)
  - then follow absolute rules
- absolute
  - strip out repeated `/`s
  - strip out unnecessary `.`s
  - strip out `dir/..`
- dir
  - strip trailing `/`
- symlink
  - are NOT followed

**compressed**

- tidy local filesystem path

**maildir**

- tidy local filesystem path

**mbox**

- tidy local filesystem path

**mh**

- tidy local filesystem path

**mmdf**

- tidy local filesystem path

**imap**

- lowercase scheme
- strip trailing `/`
- replace empty path with INBOX

**nntp**

- lowercase scheme

**notmuch**

- lowercase scheme
- tidy database path

**pop**

- lowercase scheme
- replace empty path with INBOX

## CANON - Rules for creating a canonical Mailbox Path

**Local filesystem paths**:

- absolute
  - follow symlinks

**compressed**

- canon local filesystem path

**maildir**

- canon local filesystem path

**mbox**

- canon local filesystem path

**mh**

- canon local filesystem path

**mmdf**

- canon local filesystem path

**imap**

- nothing to do

**nntp**

- nothing to do

**pop**

- nothing to do

**notmuch**

- parse URI into: `{ path, queries }`
- sort queries
- unique queries

## COMPARE - Rules for comparing Mailbox Paths

**compressed**

- canon path: exact string match

**maildir**

- canon path: exact string match

**mbox**

- canon path: exact string match

**mh**

- canon path: exact string match

**mmdf**

- canon path: exact string match

**imap**

- parse URI into: `{ user, pass, host, port, path }`
- `host` must match
- `user` must match, or may be absent from one, or absent from both
- `pass` must match, or may be absent from one, or absent from both
- `port` must match, or may be absent from one, or absent from both
- `path` must match

**nntp**

- parse URI into: `{ user, pass, host, port, path }`
- `host` must match
- `user` must match, or may be absent from one, or absent from both
- `pass` must match, or may be absent from one, or absent from both
- `port` must match, or may be absent from one, or absent from both
- `path` must match

**pop**

- parse URI into: `{ user, pass, host, port }`
- `host` must match
- `user` must match, or may be absent from one, or absent from both
- `pass` must match, or may be absent from one, or absent from both
- `port` must match, or may be absent from one, or absent from both

**notmuch**

- parse URI into: `{ path, queries }`
- `path` must match
- `sort` queries
- `queries` must match

## PARENT - Rules for finding the parent of a Mailbox Path

**compressed**

- not relevant

**notmuch**

- not relevant

**pop**

- not relevant

**mbox**

- not relevant

**mmdf**

- not relevant

**imap**

- INBOX has no parent
- parse path using server, or user, path delimiters
- strip one component from the end
- replace empty path with INBOX

**nntp**

- parse path using `.` as a separator
- strip one component from the end
- check if new path exists

**maildir**

- parse path using `/` as a separator
- strip one component from the end
- check if new path is a maildir

**mh**

- parse path using `/` as a separator
- strip one component from the end
- check if new path is a mh

## PRETTY - Rules for abbreviating a Mailbox Path

**Local filesystem paths**:

- simple string comparison of path against `$folder`
- return `+remainder`

**compressed**

- pretty local filesystem path

**maildir**

- pretty local filesystem path

**mbox**

- pretty local filesystem path

**mh**

- pretty local filesystem path

**mmdf**

- pretty local filesystem path

**pop**

- not relevant

**notmuch**

- not relevant

**imap**

- use compare rules to match `{ user, pass, host, port }`
- then abbreviate path

**nntp**

- use compare rules to match `{ user, pass, host, port }`
- then abbreviate path

**Comment by flatcap** [comment]
I've tried to trace all the points in NeoMutt where the user enters a Mailbox Path.
It turns out there are more than I imagined.

**Config Variables**
- `$folder`
- `$mbox`
- `$postponed`
- `$record`
- `$spoolfile`
- `$trash`

**Commands**:
- `fcc-hook`
- `fcc-save-hook`
- `mailboxes`
- `mbox-hook`
- `save-hook`
- `sidebar_whitelist`
- `unmailboxes`
- `unsidebar_whitelist`

**Index/Pager functions**:
- `<change-folder>`
- `<change-folder-readonly>`
- `<change-newsgroup>`
- `<change-newsgroup-readonly>`
- `<copy-message>`
- `<decode-copy>`
- `<decode-save>`
- `<decrypt-copy>`
- `<decrypt-save>`
- `<save-message>`
- `<vfolder-from-query>`
- `<vfolder-from-query-readonly>`

**Compose functions**:
- `<attach-message>`
- `<attach-news-message>`

**Browser functions**:
- `<change-dir>`
- `<create-mailbox>`
- `<goto-parent>`
- `<rename-mailbox>`

**IMAP function**:
- cmd_parse_lsub() delegate to parse_mailboxes

---

When resolving a Mailbox Path...

**What might exist?**
- account + matching mailbox
- account + no match
- no matching account + valid mailbox
- no matching account + invalid mailbox

**What's the lifespan of the Mailbox Path?**
- used + dropped
- used + incorporated into a Mailbox

**Comment by flatcap** [comment]
## Recent changes

I've extended the `Path` to include the descriptive name.
This allows us to transparently use the descriptive name in user functions.

```C
struct Mailbox
{
  ...
  struct Path *path; ///< Path representing the Mailbox
  ...
}
```

## Walk-through

Here's a walk-through of `<save-message>`

- <kbd>s</kbd> -- `<save-message>` / `OP_SAVE`

- Ask user for a location: `L`

- Search all Mailboxes for a `M->path->desc` that matches `L`
  Priority is given to the Mailboxes on the current Account.
  If successful, go to `mbox_open()`

- Resolve `L` into a Path `P`
  Resolve NeoMutt shortcuts, such as `+`, `^`, `!`, ...
  Probe `P` to determine its Mailbox type
  Tidy `P` using the API function: `path_tidy()`

- Find an existing Mailbox by Path `P`
  Search all Mailboxes, for a match
  Use the API function `path_compare()` for comparisons
  If successful, go to `mbox_open()`

- Does the path exist?
  Use the API function `path_exists()`

- Create the path? (if necessary)
  Use the API function `path_create()`

- Create a new Mailbox `M`
  The new Mailbox wraps and takes ownership of `P`

- Open the Mailbox
  Use the API function `mbox_open()`

- Create a new Context (if necessary)
  The new Context wraps and takes ownership of `M`

**Comment by flatcap** [comment]
**Update**

- Path functions are finished (but not plugged in)
- Test functions give 100% coverage of the path functions

**Next**

- Tidy / refactor the path functions
- Tidy / refactor the new tests
- Make everything robust

**Comment by flatcap** [comment]
**Update**

I'm quite happy with the path functions now.
The API docs are now pretty verbose:
- Function prototype
- Docs for each parameter
- Retvals -- All possible returned values
- Contract -- Guarantees for the function
- Success -- Guarantees for the returns values/objecs
- Notes -- Any additional information

https://github.com/neomutt/neomutt/blob/a029102fa77d2af9fd3d4a2de3fecda62ee42f33/mx.h#L312-L336

The tests:
- Cover 100% of the path code
- Check the results against the 'Success' guarantee

https://github.com/neomutt/neomutt/blob/a029102fa77d2af9fd3d4a2de3fecda62ee42f33/test/mx_path/imap.c#L136-L147

Currently, Travis is failing because some of the tests require local files.
I've put sample files in a new repo: [neomutt-test-files](https://github.com/neomutt/neomutt-test-files)
Unfortunately, they're hard-coded to `/home/mutt/path` -- suggestions welcome.

**Comment by flatcap** [comment]
No changes, just split the commits into manageable chunks.
(note: the first 5 commits are #2137)

File: p1 [hunk 1]
```diff
@@ -0,0 +1,69 @@
+command_parse.c:1366:18: mutt_pretty_mailbox(buf->data, buf->dsize);
```
Comment (andir) [inline p1]: I guess this did sneak in by accident?

Comment (flatcap) [reply p1]: er, sort of.
The whole PR is still a work-in-progress and `p1` are notes of what I was doing last.
I have so many tiny temporary files like this, that I've started committing them so I don't lose them.

**Comment by flatcap** [comment]
Archived to https://github.com/neomutt/neomutt-old#neomutt-2021-02-05

**Pull request closed without merging**